src/commons/database-operations.service.ts

```
@@ -0,0 +1,197 @@
+      this.loggerService.log(`Saved transaction history with key: ${key}`);
+    } catch (error) {
+      this.handleDatabaseError(error, 'save transaction history', {
+        details: `key ${key}, table ${destination}`,
+      });
+    }
+  }
+
+  async saveTransactionRelationship(relationship: TransactionRelationship): Promise<void> {
+    try {
+      await this.databaseService.query('INSERT INTO transaction (source, destination, transaction) VALUES ($1, $2, $3)', [
+        relationship.from,
+        relationship.to,
+        relationship,
+      ]);
+
+      this.loggerService.log(`Saved transaction relationship: ${relationship.from} -> ${relationship.to}`);
+    } catch (error) {
+      // Special validation before handling database error
+      if (!relationship.from || !relationship.to) {
+        this.loggerService.error(`Missing required fields in transaction relationship: from=${relationship.from}, to=${relationship.to}`);
+        throw new BadRequestException('Transaction relationship must have both source and destination');
+      }
+
+      this.handleDatabaseError(error, 'save transaction relationship', {
+        details: `${relationship.from} -> ${relationship.to}`,
+      });
+    }
+  }
+
+  async saveToQuarantine(payload: any, endpoint: string, differences: string[], tenantId: string, correlationId?: string): Promise<void> {
+    try {
+      const quarantineRecord = {
+        id: randomUUID(),
+        correlation_id: correlationId || null,
+        tenant_id: tenantId,
+        endpoint_path: endpoint,
+        config_id: null,
+        version: null,
+        error: JSON.stringify({
+          code: 'VALIDATION_ERROR',
+          message: 'Payload validation failed',
+          differences: differences,
+          timestamp: new Date().toISOString(),
+        }),
+        raw_payload: JSON.stringify(payload),
+        status: 'failed',
```

create status enum instead of hardcoded value "failed" here.